### PR TITLE
Add dump_format=json option

### PR DESCRIPTION
**Pull request by AbdealiJK**
This adds the option to dump as JSON.
The bindings for python, R, and java have hardcoded the format as "text" to retain the earlier usage.

Related issue is at https://github.com/dmlc/xgboost/issues/1722

### CODE REVIEW

**Comment by AbdealiJK** [comment]
Um, what's up with Travis ? The R build's bootstrap seems to be failing when downloading R

**Comment by AbdealiJK** [comment]
Here's an example of the JSON it produces with the mushroom example:
```json
[
  { "nodeid": 0, "split": "odor=pungent", "yes": 2, "no": 1, "children": [
    { "nodeid": 1, "split": "stalk-root=cup", "yes": 4, "no": 3, "children": [
      { "nodeid": 3, "split": "stalk-root=missing", "yes": 8, "no": 7, "children": [
        { "nodeid": 7, "leaf": 1.899 },
        { "nodeid": 8, "leaf": -1.94737 }
      ]},
      { "nodeid": 4, "split": "bruises?=no", "yes": 10, "no": 9, "children": [
        { "nodeid": 9, "leaf": 1.78378 },
        { "nodeid": 10, "leaf": -1.98135 }
      ]}
    ]},
    { "nodeid": 2, "split": "spore-print-color=orange", "yes": 6, "no": 5, "children": [
      { "nodeid": 5, "split": "stalk-surface-below-ring=silky", "yes": 12, "no": 11, "children": [
        { "nodeid": 11, "leaf": -1.98546 },
        { "nodeid": 12, "leaf": 0.938776 }
      ]},
      { "nodeid": 6, "leaf": 1.87097 }
    ]}
  ]},
  { "nodeid": 0, "split": "odor=pungent", "yes": 2, "no": 1, "children": [
    { "nodeid": 1, "split": "bruises?=no", "yes": 4, "no": 3, "children": [
      { "nodeid": 3, "leaf": 1.14608 },
      { "nodeid": 4, "split": "gill-spacing=crowded", "yes": 8, "no": 7, "children": [
        { "nodeid": 7, "leaf": -6.87995 },
        { "nodeid": 8, "leaf": -0.106592 }
      ]}
    ]},
    { "nodeid": 2, "split": "spore-print-color=orange", "yes": 6, "no": 5, "children": [
      { "nodeid": 5, "split": "gill-size=narrow", "yes": 10, "no": 9, "children": [
        { "nodeid": 9, "leaf": -0.0930658 },
        { "nodeid": 10, "leaf": -1.15261 }
      ]},
      { "nodeid": 6, "leaf": 1.00423 }
    ]}
  ]}
]
```

**Comment by tqchen** [comment]
Thanks for the contribution. For the sake of backward compatiblity. Let us not change the DumpModel function, but instead create a new funciton DumpModelEx, with the additional option. and redirect the legacy dump function to it.

**Comment by AbdealiJK** [comment]
That makes sense, will do

**Comment by AbdealiJK** [comment]
Done. I've added a new DumpModelEx function in the c_api and the bindings continue to use that.

**Comment by tqchen** [comment]
Please update a python wrapper, and add a testcase, then we can merge it in

**Comment by AbdealiJK** [comment]
Could you elaborate on what you mean by updating the python wrapper? Do you
mean add a new function to the python wrapper to be able to dump JSON? Or
use JSON instead of the parser currently being used in get_scores()?

If you'd like to have a real-time conversation for clarity, maybe we can
discuss on Gitter.

**Comment by tqchen** [comment]
enhance getdump https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/core.py#L1041 to enable a format option

**Comment by AbdealiJK** [comment]
@tqchen Done

**Comment by tqchen** [comment]
Thanks this is merged!

**Comment by tonydifranco** [comment]
this is a great addition @AbdealiJK!  are you able to update the R wrapper as well?

**Comment by AbdealiJK** [comment]
@tonydifranco I've never used the R wrapper, so if you could add that in a PR - it would be good. Else I'll look into it when I'm more free and have time to experiment ?

**Comment by antoinecarme** [comment]
Hi all,

Thanks a lot for this feature (dump a model as JSON). Very useful.

Could you please tell if it is possible to increase the precision of the floats that are dumped (getting a maximal one is awesome ;) ?

Thanks in advance

Antoine

**Comment by AbdealiJK** [comment]
Currently it uses the same formatting that the text format uses - 6 digits. As the data type used is float for most stats in XGBoost 6 is already the max precision I think

**Comment by antoinecarme** [comment]
@AbdealiJK thanks for your feedback.

I am trying to translate the prediction of xgboost models into a SQL code, like this

https://github.com/antoinecarme/sklearn2sql-demo/blob/master/sample_outputs_round_8/XGBClassifier/iris/oracle/demo2_XGBClassifier_oracle.sql

and the loss of precision makes that I have small differences between python version and SQL version. 

I really need to get something close to a serialized form (the maximum precision). This is just an extra parameter to be added (to be confirmed).

Should I add an issue for this task ? I know of a developer who may be interested in implementing it (myself ;)

Cheers

Antoine

**Comment by khotilov** [comment]
@antoinecarme you may use a quick hack I've suggested in #1204 as a temporary workaround.

@AbdealiJK the default iostream precision (regardless of whether it's text or json format) is often not sufficient for reasonable conversions b/w decimal and binary.

This issue was brought up a few times, e.g., in addition to #1204 see also #1750, #2077. In #2268 I've proposed a mechanism allowing to easily parameterize DumpModel (including the precision), but I haven't heard any feedback from @tqchen .

**Comment by antoinecarme** [comment]
@khotilov

This hack is what I did locally. I am developping a tool tha translates xgboost models in SQL code by working on the dump result. The hack cannot however be distrivbuted without a merge ;). I can not ask all users to change xgboost this way !!!

Thanks for pointing all these issues.

1. Score (prediction) calculated by xgboost does not match what is expected from tree model generated by xgboost ==> #1204
2. xgb.dump needs argument for number of digits ===> #1750
3. Dumpmodel has insufficient precision for dumping floats ==> #2077
4. Refactor passing of parameters to DumpModel ==> #2268

The most generic and useful ones to implement are IMHO, the #2268 and #1750

There is a need for a higher-level decision (@tqchen ?) to define the priorities here.
